tests/test_proxy.py

```diff
@@ -1,29 +1,289 @@
+    assert_etcdctl_put(frontend_backend_path, backend_alias, expected_rv)
+    assert_etcdctl_put(frontend_rule_path, rule, expected_rv)
+
+
+def check_route_with_etcdl(routespec, target, data, test_deletion=False):
+    jupyterhub_routespec = traefik_utils.jupyterhub_prefix + routespec
+    backend_alias = traefik_utils.create_backend_alias_from_url(target)
+    backend_url_path = traefik_utils.create_backend_url_path(backend_alias)
+    backend_alias = traefik_utils.create_backend_alias_from_url(target)
+    backend_weight_path = traefik_utils.create_backend_weight_path(backend_alias)
+    frontend_alias = traefik_utils.create_frontend_alias_from_url(target)
+    frontend_backend_path = traefik_utils.create_frontend_backend_path(frontend_alias)
+    frontend_rule_path = traefik_utils.create_frontend_rule_path(frontend_alias)
+    if routespec.startswith("/"):
+        # Path-based route, e.g. /proxy/path/
+        rule = "PathPrefix:" + routespec
+    else:
+        # Host-based routing, e.g. host.tld/proxy/path/
+        host, path_prefix = routespec.split("/", 1)
+        path_prefix = "/" + path_prefix
+        rule = "Host:" + host + ";PathPrefix:" + path_prefix
+
+    if test_deletion:
+        expected_rv = ""
+    else:
+        expected_rv = jupyterhub_routespec + "\n" + target
+
+    # Test that (routespec, target) pair has been added to etcd
+    assert_etcdctl_get(jupyterhub_routespec, expected_rv)
+
+    # Test that (target, data) pair has been added to etcd
+    if not test_deletion:
+        expected_rv = target + "\n" + json.dumps(data)
+    assert_etcdctl_get(target, expected_rv)
+
+    # Test that a backend has been added to etcd for this target
+    if not test_deletion:
+        expected_rv = backend_url_path + "\n" + target
+    assert_etcdctl_get(backend_url_path, expected_rv)
+
+    # Test that a backend weight has been added to etcd for this target
+    if not test_deletion:
+        expected_rv = backend_weight_path + "\n" + default_backend_weight
+    assert_etcdctl_get(backend_weight_path, expected_rv)
+
+    # Test that a frontend has been added for the prev backend
+    if not test_deletion:
+        expected_rv = frontend_backend_path + "\n" + backend_alias
+    assert_etcdctl_get(frontend_backend_path, expected_rv)
+
+    # Test that a path-routing rule has been added for this frontend
+    if not test_deletion:
+        expected_rv = frontend_rule_path + "\n" + rule
+    assert_etcdctl_get(frontend_rule_path, expected_rv)
+
+
+@pytest.mark.parametrize(
+    "routespec, target, data",
+    [
+        ("/proxy/path", "http://127.0.0.1:99", {"test": "test1"}),  # Path-based routing
+        ("/proxy/path", "http://127.0.0.1:99", {}),  # Path-based routing, no data dict
+        # Host-based routing
+        ("host/proxy/path", "http://127.0.0.1:99", {"test": "test2"}),
+    ],
+)
+async def test_add_path_based_route_to_etcd(proxy, routespec, target, data):
```

Comment:
These async tests need `@pytest.mark.async` in order to run asynchronously

Reply:
There is a `pytestmark = pytest.mark.asyncio` at the beginning of the test file from the initial project skeleton which I think it's enough to make all test coroutines be run async, right?.

Reply:
oops, yes. You are absolutely right!